### PR TITLE
Rebuild validator engine to improve error messages

### PR DESCRIPTION
**Pull request by TonyFNZ**
This change includes two major changes which significantly improve the usefulness of errors reported by the linter:
- Recursively finds most relevant validation errors
- Uses YAML-js to more accurately find error ranges

### Example 1: Invalid Types
```yaml
swagger: "2.0"
info:
  title: "Sample 1"
  version: "1.0.0"
paths:
  /todos:
    get:
      description: "Returns all todo items"
      responses:
        200:
          description: "A list of todo items"
          schema:
            type: array
            items:
              type: object
              properties:
                id:
                  type: strin
                  format: uuid
                name:
                  type: string
```
| Now | Sample 1 Errors After |
---|---
| Before | Sample 1 Errors Before |

### Example 2: Invalid Properties
```yaml
swagger: "2.0"
info:
  title: "Sample 2"
  version: "1.0.0"
paths:
  /todos:
    get:
      description: "Returns all todo items"
      responses:
        200:
          description: "A list of todo items"
          schema:
            type: array
            items:
              type: object
              properties:
                id:
                  type: string
                  forma: uuid
                name:
                  type: string
```
| Now | Sample 2 Errors After |
---|---
| Before | Sample 2 Errors Before |

### CODE REVIEW

File: lib/error-util.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
+'use babel';
+
+import _ from 'lodash';
+
+
+// determine if A starts with B
+function startsWith( a, b ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: Just use [`String.prototype.startsWith()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith).

File: lib/error-util.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
+'use babel';
+
+import _ from 'lodash';
```
Comment (Arcanemagus) [inline lib/error-util.js]: I'm _very_ much against bringing in `lodash` unless absolutely necessary, and with modern JS is is _extremely_ rare that that is the case...

File: lib/error-util.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  for ( let i = 0; i < b.length; i++ ) {
+    if ( b[ i ] !== a[ i ] ) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// determine whether A ends with B
+function endsWith( a, b ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: Same story, use [`String.prototype.endsWith()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith).

File: lib/error-util.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  let c = child.path;
+
+  if ( endsWith( p, c ) ) {
+    child.path = p; // eslint-disable-line
+  } else if ( !startsWith( c, p ) ) {
+    // Handle case where child path sometimes overlaps with the parent path by one item
+    if ( p[ p.length - 1 ] === c[ 0 ] ) {
+      c = _.slice( c, 1 );
+    }
+    child.path = _.concat( p, c ); // eslint-disable-line
```
Comment (Arcanemagus) [inline lib/error-util.js]: Just use [`Array.prototype.concat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat).

File: lib/error-util.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+function updateChildPathIfRequired( parent, child ) {
+  const p = parent.path;
+  let c = child.path;
+
+  if ( endsWith( p, c ) ) {
+    child.path = p; // eslint-disable-line
+  } else if ( !startsWith( c, p ) ) {
+    // Handle case where child path sometimes overlaps with the parent path by one item
+    if ( p[ p.length - 1 ] === c[ 0 ] ) {
+      c = _.slice( c, 1 );
```
Comment (Arcanemagus) [inline lib/error-util.js]: Just use [`Array.prototype.slice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice).

File: lib/error-util.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  return true;
+}
+
+
+/**
+ * Child errors sometimes have their paths defined relative
+ * to the parent path, or overlapping with the parent path.
+ * Here we update the child path to always be full depth
+ */
+function updateChildPathIfRequired( parent, child ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: I haven't ran this code yet to be sure, but it sounds like you are simply trying to replicate [`path.resolve()`](https://nodejs.org/api/path.html#path_path_resolve_paths). If that's the case, just use that.

File: lib/error-util.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  const p = parent.path;
+  let c = child.path;
+
+  if ( endsWith( p, c ) ) {
+    child.path = p; // eslint-disable-line
```
Comment (Arcanemagus) [inline lib/error-util.js]: _Never_ disable the entire line, if you absolutely can't work around what ESLint is yelling at you for here, then disable the specific rule and provide justification.

In this case I'm guessing you are running into [`no-param-reassign`](http://eslint.org/docs/rules/no-param-reassign) which I would much rather you simply refactor the code to work with.

File: lib/error-util.js [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    }
+    child.path = _.concat( p, c ); // eslint-disable-line
+  }
+}
+/**
+ * Recursively traverses the error tree to find leaf nodes
+ */
+function findCauseErrors( error, causes ) {
+  if ( error.inner ) {
+    _.forEach( error.inner, ( child ) => {
```
Comment (Arcanemagus) [inline lib/error-util.js]: Again, there is no need for the `lodash` function, in this case I'm guessing you should be using this:

```js
Object.keys(error.inner).forEach((childID) => {
  const child = error.inner[childID];
```

(I don't think the Node.js built into table Atom includes [`Object.values()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values) yet, if it does though that would solve the need for defining a child object reference manually.)

File: lib/error-util.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    }
+  }
+  return true;
+}
+
+/**
+ * Elimiate errors which are on the path of another deeper error
+ * These are generally secondary errors which are solved by solving the deeper error
+ */
+function removeSubsetErrors( input ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: I'm not familiar with the error tree myself so I'm not sure if this is a good idea, but would people maybe be interested in these errors? If so, maybe an option is in order to enable showing these?

File: lib/error-util.js [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+ * Tweak error paths to help ensure the most relevant
+ * document element is highlighted
+ */
+function massageErrorPaths( errors ) {
+  _.forEach( errors, ( error ) => {
+    if ( error.code === 'OBJECT_ADDITIONAL_PROPERTIES' ) {
+      error.path.push( error.params[ 0 ][ 0 ] );
+    }
+
+    // if ( error.code === 'ENUM_MISMATCH' ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: Please remove this commented code, no reason to bring it in 😉.

File: lib/error-util.js [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+  return output;
+}
+
+/**
+ * Tweak error paths to help ensure the most relevant
+ * document element is highlighted
+ */
+function massageErrorPaths( errors ) {
+  _.forEach( errors, ( error ) => {
```
Comment (Arcanemagus) [inline lib/error-util.js]: See my previous comment about replacing `_.forEach()`.

File: lib/error-util.js [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  if ( !errObj.details ) {
+    return [ {
+      message: errObj.message
+    } ];
+  }
+
+  let errors = [];
+
+  // Find root cause errors
+  _.forEach( errObj.details, ( error ) => {
```
Comment (Arcanemagus) [inline lib/error-util.js]: See my previous comment about replacing `_.forEach()`.

File: lib/error-util.js [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    findCauseErrors( error, errors );
+  } );
+
+  // Remove duplicate errors
+  errors = _.uniqWith( errors, _.isEqual );
```
Comment (Arcanemagus) [inline lib/error-util.js]: First of all, since you are just checking for simple equality, why aren't you using `_.uniq()`? And as for using this in the first place, you could just use `[...new Set(a)];` (stolen from [here](http://stackoverflow.com/questions/13486479/how-to-get-an-array-of-unique-values-from-an-array-containing-duplicates-in-java)).

File: lib/error-util.js [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+  massageErrorPaths( errors );
+
+  // Remove subset errors
+  errors = removeSubsetErrors( errors );
+
+  return errors;
+}
+
+export { parseSwaggerErrors as default };
```
Comment (Arcanemagus) [inline lib/error-util.js]: Just move this to the function definition.

File: lib/range-util.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  if ( path.length ) {
+    return undefined;
+  }
+
+  return [
+    [
+      current.start_mark.line,
+      current.start_mark.column
+    ],
+    [
```
Comment (Arcanemagus) [inline lib/range-util.js]: Will the end point _always_ be defined? It's extremely rare for this to be the case, most of the time when a linter even supports this in the first place it isn't on all results.

File: lib/error-util.js (hunk 1 above)
Comment (Arcanemagus) [inline lib/error-util.js]: Please remove all the extra spaces in all function calls, I'm not sure why ESLint wasn't blowing up at you for these, as far as I knew it should be...

File: lib/validator.js [hunk 16]
```diff
@@ -1,116 +1,46 @@
...
-      spec: true,
-    },
-  };
-
   try {
-    await SwaggerParser.validate(path, swaggerParserOpts);
+    await SwaggerParser.validate( path, { validate: { spec: true } } );
     return [];
-  } catch (err) {
-    if (editor.getText() !== text) {
```
Comment (Arcanemagus) [inline lib/validator.js]: This code wasn't just in there for fun, this is to check against a real race condition where the user has changed the contents of the file between when the lint was requested and when the base linter actually provided results.

File: lib/validator.js [hunk 17]
```diff
@@ -1,116 +1,46 @@
...
-          type: 'Error',
-          text: innerDetail.message,
-          filePath: path,
-          range: extractRange(innerDetail.path, editor),
-        }
-      )).valueOf();
-      return errors;
-    }
-    return [{
+  return errors.map( ( error ) => {
```
Comment (Arcanemagus) [inline lib/validator.js]: This can be simplified to the following:

```js
return errors.map(error => ({
    type: 'Error',
    text: error.message,
    filePath: path,
    range: error.range
  })
);
```

File: spec/fixtures/sample3.yaml [hunk 18]
```diff
@@ -10,7 +10,7 @@ paths:
         default:
           description:  hello world
           schema:
-            anyOf:                  # <--- "anyOf" is not supported by Swagger 2.0.  Only "allOf" is supported.
```
Comment (Arcanemagus) [inline spec/fixtures/sample3.yaml]: Why did you remove this comment and the source link?

File: lib/error-util.js (hunk 9 above)
Comment (Arcanemagus) [inline lib/error-util.js]: I'm guessing that this is the main cause of the improvements in your before/after pictures, if that's the case it's probably just fine to go ahead with this.

File: lib/error-util.js (hunk 2 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Agreed. Will remove lodash

File: lib/error-util.js (hunk 1 above)
Comment (TonyFNZ) [inline lib/error-util.js]: This function determines if a `array` starts with another `array`.  String.prototype.startsWith() cannot be used here and there is no equivalent Array.prototype.startsWith()

Comment (TonyFNZ) [inline lib/error-util.js]: Linting issues have been fixed, sorry!

File: lib/error-util.js (hunk 3 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Same as for startsWith() above, this function is for arrays

File: lib/error-util.js (hunk 6 above)
Comment (TonyFNZ) [inline lib/error-util.js]: This code is for paths within a document tree structure, not file system paths.  path.resolve() will not work here

File: lib/error-util.js (hunk 7 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will do

File: lib/error-util.js (hunk 5 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will remove lodash

File: lib/error-util.js (hunk 4 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will remove lodash

File: lib/error-util.js (hunk 11 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will remove lodash

File: lib/error-util.js (hunk 12 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will remove lodash

File: lib/error-util.js (hunk 10 above)
Comment (TonyFNZ) [inline lib/error-util.js]: My bad

File: lib/range-util.js (hunk 15 above)
Comment (TonyFNZ) [inline lib/range-util.js]: The `yaml-js` module always returns an end marker.  The code here is a close adaptation of [the code used by the online Swagger Editor project](https://github.com/swagger-api/swagger-editor/blob/4f1bd208e5af1e5810c4d7360d4fc2ac11a2069d/scripts/services/ast-manager.js#L88)

File: lib/validator.js (hunk 16 above)
Comment (TonyFNZ) [inline lib/validator.js]: Will restore

File: lib/error-util.js (hunk 9 above)
Comment (TonyFNZ) [inline lib/error-util.js]: I write a lot of swagger files personally, and in my experience the errors hidden by this code are purely noise.

File: lib/error-util.js (hunk 8 above)
Comment (TonyFNZ) [inline lib/error-util.js]: Will remove lodash

File: lib/validator.js (hunk 17 above)
Comment (TonyFNZ) [inline lib/validator.js]: Linting fixed

File: spec/fixtures/sample3.yaml (hunk 18 above)
Comment (TonyFNZ) [inline spec/fixtures/sample3.yaml]: Was lost during conversion to JSON and back again.  Can restore if you prefer it

**Comment by TonyFNZ** [comment]
Thanks for the quick review.  On the subject of `yaml-js`, the addition of this module is purely to allow accurate identification of the range for each swagger error (not for linting the JSON/YAML format itself.

This is required as the current implementation of `extractRange()` gives the wrong result for:
- Errors which exist inside an array
- Nested paths with the same keys (eg. obj.type.type='user')

I looked at a variety of options for this problem, and `yaml-js` is used by http://editor.swagger.io/, so it felt like a good choice.

**Comment by TonyFNZ** [comment]
I believe I have addressed all the major issues now.  bitHound is showing a lint error still, but I can't figure out what the issue is.  There are no lint issues for me here, and when I click to see what the issue is bitHound says "Oops! Can't quite do that!"

**Comment by spoike** [comment]
Looked in bithound, apparently it hadn't analyzed the master branch in your fork so it couldn't analyze this branch.

It appears the error is caused by a [breaking change with airbnb eslint config](https://github.com/airbnb/javascript/commit/f6dab799b8fa1346f1cc3fdfdbe9c9fbeef7baf6) with [import/extensions rule](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md) that makes sure you use file extensions in import paths. This however totally disregards installed modules in `node_modules`.

There is a workaround with [reverting the old behavior](https://github.com/clayne11/eslint-import-resolver-meteor/issues/17#issuecomment-257091233).

**Comment by TonyFNZ** [comment]
Awesome, thanks @spoike. This should be fixed now

File: lib/error-util.js [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+
+  for (let i = 1; i <= b.length; i += 1) {
+    if (b[b.length - i] !== a[a.length - i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+
```
Comment (Arcanemagus) [inline lib/error-util.js]: Extra line between these functions.

File: lib/error-util.js [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+  if (arrayEndsWith(p, c)) {
+    child.path = p; // eslint-disable-line no-param-reassign
+  } else if (!arrayStartsWith(c, p)) {
+    // Handle case where child path sometimes overlaps with the parent path by one item
+    if (p[p.length - 1] === c[0]) {
+      c = c.slice(1);
+    }
+    child.path = p.concat(c); // eslint-disable-line no-param-reassign
+  }
+}
```
Comment (Arcanemagus) [inline lib/error-util.js]: Missing line between these functions.

File: lib/error-util.js [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+ * Tweak error paths to help ensure the most relevant
+ * document element is highlighted
+ */
+function massageErrorPaths(errors) {
+  errors.forEach((error) => {
+    if (error.code === 'OBJECT_ADDITIONAL_PROPERTIES') {
+      error.path.push(error.params[0][0]);
+    }
+
+    // if ( error.code === 'ENUM_MISMATCH' ) {
```
Comment (Arcanemagus) [inline lib/error-util.js]: Is there a reason this commented code is still here?

File: lib/error-util.js [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+    //   error.path.push( error.params[ 0 ] );
+    // }
+  });
+}
+
+
```
Comment (Arcanemagus) [inline lib/error-util.js]: Extra space here.

File: lib/validator.js [hunk 23]
```diff
@@ -1,116 +1,44 @@
...
-    },
-  };
-
   try {
-    await SwaggerParser.validate(path, swaggerParserOpts);
+    await SwaggerParser.validate(path, { validate: { spec: true } });
     return [];
   } catch (err) {
     if (editor.getText() !== text) {
-      // Editor contents have changed, tell Linter not to update messages
```
Comment (Arcanemagus) [inline lib/validator.js]: Looks like this snuck in at some point? 😛

That space should be there.

File: spec/json-files-spec.js [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+
+
+describe('Linting JSON files', () => {
+  beforeEach(() => {
+    atom.workspace.destroyActivePaneItem();
+    waitsForPromise(() => Promise.all([
+      atom.packages.activatePackage('linter-swagger'),
+      atom.packages.activatePackage('language-json'),
+    ]),
+      );
```
Comment (Arcanemagus) [inline spec/json-files-spec.js]: This should be moved to the above line.

**Comment by Arcanemagus** [comment]
Sorry I missed getting back to this till now!

**Comment by TonyFNZ** [comment]
@Arcanemagus Thanks for the re-review.  Should be all sorted now.

**Comment by Arcanemagus** [comment]
@spoike Any changes you would like made before this is merged?

**Comment by Arcanemagus** [comment]
I'm going to go with there being no further changes and get this merged 😛.

**Comment by Arcanemagus** [comment]
Published in v0.4.0 🎉

@TonyFNZ would you be interested in helping to maintain this package?
